### PR TITLE
Default header policy

### PR DESCRIPTION
**Pull request by erwbgy**
Second part of #3258

Add a policy section to the Contour configuration file with request-headers and response-headers fields that allow setting or removing default request and response headers unless overridden by users in their HTTPProxy objects.

Signed-off-by: Keith Burdis <keith.burdis@gs.com>

### CODE REVIEW

**Review by youngnick — COMMENTED** [review]
This LGTM once #3269 is merged.

**Review by sunjayBhatia — COMMENTED** [review]
Didn't see any specific discussion on this question from the original issue:

```
What happens if someone defines a header policy in their HTTPProxy that conflicts with the globally-set one? That is, what happens if someone tries to set the same header, or tries to unset the default one?
```

but looks like we have chosen to do the following:

```
The default header values override anything users set or unset in their HTTPProxy objects - the admin who owns Contour wins.
```

I could see this being a little counterintuitive just because usually a global default is used as a fallback and more specific localized config is used to override global config. Regardless, this seems like something to confirm before merging

File: pkg/config/parameters.go
Comment (sunjayBhatia) [inline pkg/config/parameters.go:332]: Could we do something like https://github.com/projectcontour/contour/blob/0829d51420663511f0c7de3d0beacc753001605c/apis/projectcontour/v1/httpproxy.go#L615-L625

or even `map[string]string` instead so we don't have to do splitting etc.

Would also reduce duplication we have here: https://github.com/projectcontour/contour/pull/3270/files#diff-1e10012fb1102e991322c6675c2ce495873c971f5c957ea170e2cf95640f8157R274

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]

**Comment by stevesloka** [comment]
> I could see this being a little counterintuitive just because usually a global default is used as a fallback and more specific localized config is used to override global config

I agree with @sunjayBhatia, many things in Contour are the opposite today. Auth, TLS redirect, etc which work in the opposite way, but I do see the value in an operator defining things that need to be applied and users shouldn't be able to "undo".

**Comment by youngnick** [comment]
For other things like this, we've:
- allowed the setting of a global default that can be overridden by the HTTPProxy owner (app dev).
- Provided Gatekeeper templates for preventing those fields from being set (so that the cluster operator can prevent them being overridden).

Doing it that way keeps it more understandable for the less strict use case, but allows for the more strict use case to work, as long as the more strict cluster runs Gatekeeper.

**Comment by erwbgy** [comment]
Ok, changed it so that request headers from the config are set _before_ the headers policy in the object is processed so that global values do not override headers set or removed in the object.

File: pkg/config/parameters.go
Comment (erwbgy) [reply pkg/config/parameters.go:332]: Looking at the existing values in the config file, I was thinking that:
```
set-request-headers:
- X-Envoy-Hostname=%HOSTNAME%
- l5d-dst-override=%CONTOUR_SERVICE_NAME%.%CONTOUR_NAMESPACE%.svc.cluster.local:%CONTOUR_SERVICE_PORT%
```
would be more natural than:
```
set-request-headers:
- name: X-Envoy-Hostname
  value: %HOSTNAME%
- name: l5d-dst-override
  value: %CONTOUR_SERVICE_NAME%.%CONTOUR_NAMESPACE%.svc.cluster.local:%CONTOUR_SERVICE_PORT%
```
Happy to change this if you think otherwise though.

Comment (erwbgy) [reply pkg/config/parameters.go:332]: Thinking some more about your comment about reducing duplication, perhaps we should do what you suggest but take it further and use:
```
defaults:
  requestHeadersPolicy:
    set:
    - name: X-Envoy-Hostname
      value: %HOSTNAME%
    - name: l5d-dst-override
      value: %CONTOUR_SERVICE_NAME%.%CONTOUR_NAMESPACE%.svc.cluster.local:%CONTOUR_SERVICE_PORT%
    remove:
    - Some-Header
    - Some-Other-Header
```
so configuration is the same as in the object itself.

My next proposal is going to be to add a default authorisation policy which could follow a similar pattern and reuse the existing objects:
```
defaults:
  authorization:
    extensionRef:
      name: htpasswd
      namespace: projectcontour-auth
  requestHeadersPolicy:
    set:
    - name: X-Envoy-Hostname
      value: %HOSTNAME%
    - name: l5d-dst-override
      value: %CONTOUR_SERVICE_NAME%.%CONTOUR_NAMESPACE%.svc.cluster.local:%CONTOUR_SERVICE_PORT%
```
In future I could see this being extended to rateLimitPolicy, timeoutPolicy, retryPolicy etc if requested by users.

Thoughts?

Comment (sunjayBhatia) [reply pkg/config/parameters.go:332]: that could be interesting, though I'm not sure our philosophy on the config file having a dependency on types from the CRDs/APIs, I could see some unfortunate migration issues longer term and we've also used the config file as a less strictly versioned surface area

cc @projectcontour/maintainers for more thoughts

Comment (sunjayBhatia) [reply pkg/config/parameters.go:332]: I think it makes sense to keep the config file package separate from the CRD config so any tool generating configs doesn't pull in the CRD packages as well

Comment (erwbgy) [reply pkg/config/parameters.go:332]: Yes, I agree that minimising dependencies between packages is good, though we could duplicate the types for now, which I think is what you were proposing above for HeaderValue?

Comment (sunjayBhatia) [reply pkg/config/parameters.go:332]: yep 👍🏽

Comment (erwbgy) [reply pkg/config/parameters.go:332]: Ok great, I'll rework the changes to use duplicates of the types.  And add more tests.

Comment (erwbgy) [reply pkg/config/parameters.go:332]: The contour/internal/dag package already has duplicate types so presumably those could be used since they're internal?

Comment (sunjayBhatia) [reply pkg/config/parameters.go:332]: unfortunately though it is a lot of duplication we would have to copy that type into the `pkg/config` package, `internal` can't be imported by code outside of `github.com/projectcontour/contour` so tools that would look to build contour config (e.g. contour-operator) wouldn't be able to use it

**Comment by youngnick** [comment]
That's looking pretty good to me at the moment @erwbgy. Do you want to bring this to ready for review (ie not "Draft")?

**Comment by erwbgy** [comment]
@youngnick It is failing the integration test at the moment but I haven't had the chance to dig into why.

**Comment by youngnick** [comment]
Ah yes, I see that the header test is the second one, and once it fails, everything else does too. I'll leave it with you, give us a shout if you need a hand.

**Comment by erwbgy** [comment]
Integration test now passes with no policy config - #3338 will make finding issues like this easier in future.

**Comment by erwbgy** [comment]
@sunjayBhatia @youngnick I think this should be good to review now.

File: examples/contour/01-contour-config.yaml
Comment (youngnick) [inline examples/contour/01-contour-config.yaml:109]:
```suggestion
    # Global policy settings.
```
Small nit.

**Review by youngnick — APPROVED** [review]
LGTM with one tiny nit. Thanks @erwbgy!

File: site/docs/main/configuration.md
Comment (skriss) [inline site/docs/main/configuration.md:216]: I think it'd be helpful to clarify that individual header values can be overridden in an individual proxy, but you can't override the set, i.e. an individual proxy can't remove an item from this globally-defined set.

File: site/docs/main/configuration.md
Comment (skriss) [inline site/docs/main/configuration.md:217]: similarly, it'd be helpful to clarify that this set of removes is appended to an individual proxy's list of removes, i.e. an individual proxy can't remove an item from the global list of removes.

**Review by skriss — CHANGES_REQUESTED** [review]

File: examples/contour/01-contour-config.yaml
Comment (erwbgy) [reply examples/contour/01-contour-config.yaml]: Fixed.

File: internal/dag/httpproxy_processor.go
Comment (erwbgy) [reply internal/dag/httpproxy_processor.go]: Fixed.

File: site/docs/main/configuration.md
Comment (erwbgy) [reply site/docs/main/configuration.md:228]: Agreed, added a note.

File: site/docs/main/configuration.md
Comment (erwbgy) [reply site/docs/main/configuration.md:229]: Agreed, added a note.

File: pkg/config/parameters.go
Comment (erwbgy) [reply pkg/config/parameters.go]: Good catches @skriss, fixed.

**Review by skriss — APPROVED** [review]
thanks for the updates @erwbgy, LGTM!

**Comment by skriss** [comment]
@sunjayBhatia GitHub really wants you to take another look at this :)

File: examples/contour/01-contour-config.yaml
Comment (sunjayBhatia) [inline examples/contour/01-contour-config.yaml:130]: the tiniest of nits
```suggestion
    #       X-Envoy-Hostname: %HOSTNAME%
```

File: pkg/config/parameters.go
Comment (sunjayBhatia) [inline pkg/config/parameters.go:420]: Do we think this type will be expanded upon to support more than headers? (hence the more general name?)

**Review by sunjayBhatia — APPROVED** [review]
LGTM overall just 2 small comments

File: examples/contour/01-contour-config.yaml
Comment (erwbgy) [reply examples/contour/01-contour-config.yaml]: Fixed.

File: pkg/config/parameters.go
Comment (erwbgy) [reply pkg/config/parameters.go:420]: Yes, exactly - anything with Policy in the name should be able to be configured here. My next target is authPolicy and possibly rateLimitPolicy after that.

**Comment by skriss** [comment]
@sunjayBhatia are you good here? ready to merge if so.

**Comment by sunjayBhatia** [comment]
> @sunjayBhatia are you good here? ready to merge if so.

yep sounds good to me

**Pull request merged**
